src/index.js

```diff
@@ -202,172 +286,266 @@ class PeerId {
-  pubKey = pubKey ? await cryptoKeys.unmarshalPublicKey(pubKey) : false
+    /**
+     * @type {Uint8Array}
+     */
+    let privDigest = new Uint8Array()
 
-  let pubDigest
-  let privDigest
+    if (privKey) {
+      // @ts-ignore
+      privDigest = await computeDigest(privKey.public)
+    }
 
-  if (privKey) {
-    privDigest = await computeDigest(privKey.public)
-  }
+    if (pubKey) {
+      pubDigest = await computeDigest(pubKey)
+    }
 
-  if (pubKey) {
-    pubDigest = await computeDigest(pubKey)
-  }
+    if (privKey) {
+      if (pubKey) {
+        if (!uint8ArrayEquals(privDigest, pubDigest)) {
+          throw new Error('Public and private key do not match')
+        }
+      }
+      // @ts-ignore
+      return new PeerId(privDigest, privKey, privKey.public)
+    }
 
-  if (privKey) {
+    // TODO: val id and pubDigest
     if (pubKey) {
-      if (!uint8ArrayEquals(privDigest, pubDigest)) {
-        throw new Error('Public and private key do not match')
-      }
+      return new PeerId(pubDigest, undefined, pubKey)
     }
-    return new PeerIdWithIs(privDigest, privKey, privKey.public)
-  }
 
-  // TODO: val id and pubDigest
+    if (id) {
+      return new PeerId(id)
+    }
 
-  if (pubKey) {
-    return new PeerIdWithIs(pubDigest, null, pubKey)
+    throw new Error('Protobuf did not contain any usable key material')
   }
 
-  if (id) {
-    return new PeerIdWithIs(id)
+  /**
+  * Checks if a value is an instance of PeerId.
+  * 
+  * @param {any} peerId - The value to check.
+  * @returns {boolean}
+  */
+  static isPeerId (peerId) {
+    return Boolean(typeof peerId === 'object' && peerId._id && peerId._idB58String)
+  } 
+}
+
+
+/**
+ * Compute digest.
+ * 
+ * @param {PublicKey} pubKey 
+ * @returns {Promise<Uint8Array>}
+ */
+const computeDigest = async (pubKey) => {
+  if (pubKey.bytes.length <= 42) {
+    return Promise.resolve(mh.encode(pubKey.bytes, 'identity'))
+  } else {
+    return pubKey.hash()
   }
+}
 
-  throw new Error('Protobuf did not contain any usable key material')
+/**
+ * Compute PeerId.
+ * 
+ * @param {PublicKey} pubKey 
+ * @param {PrivateKey} [privKey]
+ * @returns {Promise<PeerId>}
+ */
+const computePeerId = async (pubKey, privKey) => {
+  const digest = await computeDigest(pubKey)
+  return new PeerId(digest, privKey, pubKey)
 }
 
-exports.isPeerId = (peerId) => {
-  return Boolean(typeof peerId === 'object' &&
-    peerId._id &&
-    peerId._idB58String)
+/**
+ * Create PeerId from CID.
+ * 
+ * @param {CID | {codec: string}} cid - The CID.
+ * @returns {boolean}
+ */
+const validMulticodec = (cid) => {
+  // supported: 'libp2p-key' (CIDv1) and 'dag-pb' (CIDv0 converted to CIDv1)
+  return cid.codec === 'libp2p-key' || cid.codec === 'dag-pb'
 }
 
-function toB64Opt (val) {
+/**
+ * @param {Uint8Array | undefined} val 
+ * @returns {string}
+ */
+const toB64Opt = (val) => {
   if (val) {
     return uint8ArrayToString(val, 'base64pad')
   }
+
+  return ''
```

Comment:
I don't like returning things that are not valid. I think we just do in `toJson`:

```js
const privKey = this.marshalPrivKey()
const pubKey = this.marshalPubKey()

return {
      id: this.toB58String(),
      privKey: privKey && toB64Opt(privKey),
      pubKey: pubKey && toB64Opt(pubKey)
}
```

This way, this function could not accept undefined and this would return what is suppoed or throw